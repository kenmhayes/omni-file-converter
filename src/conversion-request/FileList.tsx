import React from 'react';
import { ListGroup } from 'react-bootstrap';
import icons from '../assets/svg/bootstrap-icons.svg';

export interface FileListProps {
  files: File[];
  setFiles: (files: File[]) => void;
}

interface FileListItemProps {
  fileName: string;
  onDelete: () => void;
}

function FileListItem(props: FileListItemProps) {
  const { fileName, onDelete } = props;

  return (
    <div className="container">
      <div className="row align-items-center">
        <div className="col-4">
          {fileName}
        </div>
        <div className="col">
          <button type="button" aria-label="remove-file" className="btn" title="Remove File" onClick={onDelete}>
            <svg className="bi" width="32" height="32" fill="currentColor">
              <use xlinkHref={`${icons}#x-circle`} />
            </svg>
          </button>
        </div>
      </div>
    </div>
  );
}

function FileList(props: FileListProps) {
  const { files, setFiles } = props;

  const onItemDelete = (index: number) => {
    const filesCopy = [...files];
    filesCopy.splice(index, 1);
    setFiles(filesCopy);
  };

  return (
    <div className="container">
      <h6>Uploaded files:</h6>
      <ListGroup variant="flush">
        {files.map(
          (file: File, index: number) => (
            <ListGroup.Item className="list-group-item" key={`${file.name}_index`}>
              <FileListItem
                fileName={file.name}
                onDelete={() => onItemDelete(index)}
              />
            </ListGroup.Item>
          ),
        )}
      </ListGroup>

    </div>
  );
}

export default FileList;
