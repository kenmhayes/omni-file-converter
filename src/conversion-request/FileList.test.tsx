import React from 'react';
import { create } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import FileList from './FileList';

it('renders a file list', () => {
  const files: File[] = [
    { name: "File 1" } as File, { name: "File 2" } as File, { name: "File 3"} as File
  ];
  const component = <FileList files={files} setFiles={jest.fn()}/>;

  expect(create(component).toJSON()).toMatchSnapshot();
});


it('update file from the list when the delete button is pressed', () => {
  const files: File[] = [
    { name: "File 1" } as File, { name: "File 2" } as File, { name: "File 3"} as File
  ];

  const setFileHandler = jest.fn();

  const { queryAllByRole, getAllByRole } = render(
    <FileList files={files} setFiles={setFileHandler}/>
  );

  const deleteButtonRole = "button";
  const deleteButtonOptions = { name: /remove-file/i };
  expect(queryAllByRole(deleteButtonRole, deleteButtonOptions)).toBeTruthy();
  let deleteButtons = queryAllByRole(deleteButtonRole, deleteButtonOptions);

  // Delete second element
  fireEvent.click(deleteButtons[1]);

  expect(setFileHandler).toHaveBeenCalledWith([files[0], files[2]]);

  // Delete first element
  deleteButtons = getAllByRole(deleteButtonRole, deleteButtonOptions);
  fireEvent.click(deleteButtons[0]);

  expect(setFileHandler).toHaveBeenCalledWith([files[1], files[2]]);
});

