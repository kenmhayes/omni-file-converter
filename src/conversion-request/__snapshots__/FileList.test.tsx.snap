// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a file list 1`] = `
<div
  className="container"
>
  <h6>
    Uploaded files:
  </h6>
  <div
    className="list-group list-group-flush"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="row align-items-center"
        >
          <div
            className="col-4"
          >
            File 1
          </div>
          <div
            className="col"
          >
            <button
              aria-label="remove-file"
              className="btn"
              onClick={[Function]}
              title="Remove File"
              type="button"
            >
              <svg
                className="bi"
                fill="currentColor"
                height="32"
                width="32"
              >
                <use
                  xlinkHref="bootstrap-icons.svg#x-circle"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="list-group-item list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="row align-items-center"
        >
          <div
            className="col-4"
          >
            File 2
          </div>
          <div
            className="col"
          >
            <button
              aria-label="remove-file"
              className="btn"
              onClick={[Function]}
              title="Remove File"
              type="button"
            >
              <svg
                className="bi"
                fill="currentColor"
                height="32"
                width="32"
              >
                <use
                  xlinkHref="bootstrap-icons.svg#x-circle"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="list-group-item list-group-item"
      data-rr-ui-event-key={null}
      onClick={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="row align-items-center"
        >
          <div
            className="col-4"
          >
            File 3
          </div>
          <div
            className="col"
          >
            <button
              aria-label="remove-file"
              className="btn"
              onClick={[Function]}
              title="Remove File"
              type="button"
            >
              <svg
                className="bi"
                fill="currentColor"
                height="32"
                width="32"
              >
                <use
                  xlinkHref="bootstrap-icons.svg#x-circle"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
